In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
pd.options.display.max_rows = 200
pd.options.display.max_columns = 200

**读取数据**

In [2]:
#读取数据
df = pd.read_csv('../data/LoanStats_2018Q1.csv', header = 1, low_memory=False)
df1 = pd.read_csv('../data/LoanStats_2018Q2.csv', header = 1, low_memory=False)

In [3]:
df = pd.concat([df,df1],ignore_index = True)
df.info(max_cols = 150)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 238640 entries, 0 to 238639
Data columns (total 145 columns):
id                                            4 non-null object
member_id                                     0 non-null float64
loan_amnt                                     238636 non-null float64
funded_amnt                                   238636 non-null float64
funded_amnt_inv                               238636 non-null float64
term                                          238636 non-null object
int_rate                                      238636 non-null object
installment                                   238636 non-null float64
grade                                         238636 non-null object
sub_grade                                     238636 non-null object
emp_title                                     218361 non-null object
emp_length                                    218740 non-null object
home_ownership                                238636 non-null obj

In [4]:
df.tail()

id  member_id  \
238635                                               NaN        NaN   
238636                                               NaN        NaN   
238637                                               NaN        NaN   
238638  Total amount funded in policy code 1: 2080429200        NaN   
238639   Total amount funded in policy code 2: 737901574        NaN   

        loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
238635    14000.0      14000.0          13975.0   36 months   16.01%   
238636     7500.0       7500.0           7500.0   36 months   10.90%   
238637    35000.0      35000.0          35000.0   36 months    6.07%   
238638        NaN          NaN              NaN         NaN      NaN   
238639        NaN          NaN              NaN         NaN      NaN   

        installment grade sub_grade            emp_title emp_length  \
238635       492.27     C        C5       client analyst  10+ years   
238636       245.19     B        B4  Residence Counselor  10+ years   
238637      1065.88     A        A2              Founder  10+ years   
238638          NaN   NaN       NaN                  NaN        NaN   
238639          NaN   NaN       NaN                  NaN        NaN   

       home_ownership  annual_inc verification_status   issue_d loan_status  \
238635       MORTGAGE     52874.0        Not Verified  Apr-2018     Current   
238636           RENT    126000.0        Not Verified  Apr-2018     Current   
238637           RENT    110000.0     Source Verified  Apr-2018     Current   
238638            NaN         NaN                 NaN       NaN         NaN   
238639            NaN         NaN                 NaN       NaN         NaN   

       pymnt_plan  url  desc             purpose               title zip_code  \
238635          n  NaN   NaN  debt_consolidation  Debt consolidation    912xx   
238636          n  NaN   NaN               other               Other    125xx   
238637          n  NaN   NaN  debt_consolidation  Debt consolidation    341xx   
238638        NaN  NaN   NaN                 NaN                 NaN      NaN   
238639        NaN  NaN   NaN                 NaN                 NaN      NaN   

       addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
238635         CA  22.36          0.0         Sep-1995             1.0   
238636         NY   6.33          0.0         Dec-1986             1.0   
238637         FL   9.80          0.0         Jul-1995             0.0   
238638        NaN    NaN          NaN              NaN             NaN   
238639        NaN    NaN          NaN              NaN             NaN   

        mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
238635                     NaN                    87.0       9.0      1.0   
238636                     NaN                     NaN      16.0      0.0   
238637                     NaN                     NaN      13.0      0.0   
238638                     NaN                     NaN       NaN      NaN   
238639                     NaN                     NaN       NaN      NaN   

        revol_bal revol_util  total_acc initial_list_status  out_prncp  \
238635    25426.0        85%       20.0                   f   11774.07   
238636    11122.0      28.3%       31.0                   f    6226.25   
238637    39634.0      35.2%       21.0                   w   28682.91   
238638        NaN        NaN        NaN                 NaN        NaN   
238639        NaN        NaN        NaN                 NaN        NaN   

        out_prncp_inv  total_pymnt  total_pymnt_inv  total_rec_prncp  \
238635       11753.05      3433.44          3427.31          2225.93   
238636        6226.25      1707.25          1707.25          1273.75   
238637       28682.91      7449.36          7449.36          6317.09   
238638            NaN          NaN              NaN              NaN   
238639            NaN          NaN              NaN              NaN   

        total_rec_i

数据来自于 Lending Club  
https://www.lendingclub.com/info/download-data.action

In [5]:
#对于变量类型为obejct的数据进行察看
df.select_dtypes(include = ['O']).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 238640 entries, 0 to 238639
Data columns (total 38 columns):
id                           4 non-null object
term                         238636 non-null object
int_rate                     238636 non-null object
grade                        238636 non-null object
sub_grade                    238636 non-null object
emp_title                    218361 non-null object
emp_length                   218740 non-null object
home_ownership               238636 non-null object
verification_status          238636 non-null object
issue_d                      238636 non-null object
loan_status                  238636 non-null object
pymnt_plan                   238636 non-null object
purpose                      238636 non-null object
title                        238636 non-null object
zip_code                     238636 non-null object
addr_state                   238636 non-null object
earliest_cr_line             238636 non-null object
revol_util

对于评分卡建立来说，term,grade,sub_grade,issue_d,title均属于贷后数据因此决定删除。但是loan_status对于我们进行数据标记有用处，所以不删除。

In [6]:
#对于变量类型为obejct的数据进行察看
df.select_dtypes(include = ['O']).head()

id        term int_rate grade sub_grade               emp_title  \
0  NaN   60 months   16.01%     C        C5  Case Manager Associate   
1  NaN   60 months   14.07%     C        C3                 Teacher   
2  NaN   36 months   10.41%     B        B3                 Teacher   
3  NaN   36 months    7.96%     A        A5  Occupational Therapist   
4  NaN   36 months    6.07%     A        A2         Credit Director   

  emp_length home_ownership verification_status   issue_d loan_status  \
0    2 years            OWN        Not Verified  Mar-2018     Current   
1    4 years           RENT     Source Verified  Mar-2018     Current   
2  10+ years       MORTGAGE        Not Verified  Mar-2018     Current   
3    3 years       MORTGAGE        Not Verified  Mar-2018     Current   
4  10+ years            OWN     Source Verified  Mar-2018     Current   

  pymnt_plan             purpose               title zip_code addr_state  \
0          n  debt_consolidation  Debt consolidation    328xx         FL   
1          n      major_purchase      Major purchase    341xx         FL   
2          n             medical    Medical expenses    786xx         TX   
3          n               other               Other    481xx         MI   
4          n               house         Home buying    117xx         NY   

  earliest_cr_line revol_util initial_list_status last_pymnt_d next_pymnt_d  \
0         Oct-1993      36.1%                   w     Nov-2018     Dec-2018   
1         Oct-2004      15.7%                   w     Nov-2018     Dec-2018   
2         Sep-2000      21.1%                   w     Nov-2018     Dec-2018   
3         Aug-2011      74.5%                   w     Nov-2018     Dec-2018   
4         Mar-1994      11.3%                   w     Nov-2018     Dec-2018   

  last_credit_pull_d application_type verification_status_joint  \
0           Nov-2018       Individual                       NaN   
1           Nov-2018       Individual                       NaN   
2           Nov-2018       Individual                       NaN   
3           Nov-2018       Individual                       NaN   
4           Nov-2018       Individual                       NaN   

  sec_app_earliest_cr_line hardship_flag hardship_type hardship_reason  \
0                      NaN             N           NaN             NaN   
1                      NaN             N           NaN             NaN   
2                      NaN             N           NaN             NaN   
3                      NaN             N           NaN             NaN   
4                      NaN             N           NaN             NaN   

  hardship_status hardship_start_date hardship_end_date  \
0             NaN                 NaN               NaN   
1             NaN                 NaN               NaN   
2             NaN                 NaN               NaN   
3             NaN                 NaN               NaN   
4             NaN                 NaN               NaN   

  payment_plan_start_date hardship_loan_status disbursement_method  \
0                     NaN                  NaN                Cash   
1                     NaN                  NaN                Cash   
2                     NaN                  NaN                Cash   
3                     NaN                  NaN                Cash   
4                     NaN                  NaN                Cash   

  debt_settlement_flag debt_settlement_flag_date settlement_status  \
0                    N                       NaN               NaN   
1                    N                       NaN               NaN   
2                    N                       NaN               NaN   
3                    N                       NaN               NaN   
4                    N                       NaN               NaN   

  settlement_date  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN

发现：  
1.从这里可以发现，id这一栏数据只有2个非空，而其他的数据最多的只有238636行，这两个加起来正好238640也就是总的数据条数，这就说明有可能是这4条变量本身就是空的，需要察看。  
2.hardship_flag代表着困难情况，具体里面的数据代表的含义，之后进行分析。  
3.verification_status_joint,sec_app_earliest_cr_line,等变量数量太少，不进行分析。(本次分析按照65%的阈值来进行删除)。
4.emp_title,emp_length这两个字段的缺失值较少的字段需要填充。  
5.int_rate, revol_util这两个本应该为数字的字段这里是用字符串表示的，为了之后的计算方便，需要更改类型。

In [7]:
#察看id 不为空的字段的信息
df.loc[df.id.notnull()]

id  member_id  \
107864  Total amount funded in policy code 1: 1741781700        NaN   
107865   Total amount funded in policy code 2: 564202131        NaN   
238638  Total amount funded in policy code 1: 2080429200        NaN   
238639   Total amount funded in policy code 2: 737901574        NaN   

        loan_amnt  funded_amnt  funded_amnt_inv term int_rate  installment  \
107864        NaN          NaN              NaN  NaN      NaN          NaN   
107865        NaN          NaN              NaN  NaN      NaN          NaN   
238638        NaN          NaN              NaN  NaN      NaN          NaN   
238639        NaN          NaN              NaN  NaN      NaN          NaN   

       grade sub_grade emp_title emp_length home_ownership  annual_inc  \
107864   NaN       NaN       NaN        NaN            NaN         NaN   
107865   NaN       NaN       NaN        NaN            NaN         NaN   
238638   NaN       NaN       NaN        NaN            NaN         NaN   
238639   NaN       NaN       NaN        NaN            NaN         NaN   

       verification_status issue_d loan_status pymnt_plan  url  desc purpose  \
107864                 NaN     NaN         NaN        NaN  NaN   NaN     NaN   
107865                 NaN     NaN         NaN        NaN  NaN   NaN     NaN   
238638                 NaN     NaN         NaN        NaN  NaN   NaN     NaN   
238639                 NaN     NaN         NaN        NaN  NaN   NaN     NaN   

       title zip_code addr_state  dti  delinq_2yrs earliest_cr_line  \
107864   NaN      NaN        NaN  NaN          NaN              NaN   
107865   NaN      NaN        NaN  NaN          NaN              NaN   
238638   NaN      NaN        NaN  NaN          NaN              NaN   
238639   NaN      NaN        NaN  NaN          NaN              NaN   

        inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
107864             NaN                     NaN                     NaN   
107865             NaN                     NaN                     NaN   
238638             NaN                     NaN                     NaN   
238639             NaN                     NaN                     NaN   

        open_acc  pub_rec  revol_bal revol_util  total_acc  \
107864       NaN      NaN        NaN        NaN        NaN   
107865       NaN      NaN        NaN        NaN        NaN   
238638       NaN      NaN        NaN        NaN        NaN   
238639       NaN      NaN        NaN        NaN        NaN   

       initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
107864                 NaN        NaN            NaN          NaN   
107865                 NaN        NaN            NaN          NaN   
238638                 NaN        NaN            NaN          NaN   
238639                 NaN        NaN            NaN          NaN   

        total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
107864              NaN              NaN            NaN                 NaN   
107865              NaN              NaN            NaN                 NaN   
238638              NaN              NaN            NaN                 NaN   
238639              NaN              NaN            NaN                 NaN   

        recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
107864         NaN                      NaN          NaN              NaN   
107865         NaN                      NaN          NaN              NaN   
238638         NaN                      NaN          NaN              NaN   
238639         NaN                      NaN          NaN              NaN   

       next_pymnt_d last_credit_pull_d  collections_12_mths_ex_med  \
107864          NaN                NaN                         NaN   
107865          NaN                NaN                         NaN   
238638          NaN                NaN                         NaN   
238639          NaN                NaN                         NaN   

        mths_since

可以发现这4条数据中基本没有信息，因此决定删除。

In [8]:
#察看issue_d
df.issue_d.value_counts()

May-2018    46311
Apr-2018    42928
Jun-2018    41533
Mar-2018    38771
Jan-2018    36347
Feb-2018    32746
Name: issue_d, dtype: int64

符合本次数据集的时间区段的时间含义，借款发生的时间为1-6月，并且和为238636，说明没有缺失的放贷日期。不过放贷时间对于评分卡构建没有影响，因此决定删除。

In [9]:
#察看hardship字段究竟代表什么意思
df.hardship_flag.value_counts()

N    238540
Y        96
Name: hardship_flag, dtype: int64

从中可以发现，存在困难情况的只有96家，说明数据集中存在困难情况的基本没有，因此可以说明，与困难情况有关系的字段方差过小，在评分卡中毫无区分性，因此不进行分析。

In [10]:
#察看loan_status的内容
df.loan_status.value_counts()

Current               207975
Fully Paid             20419
Late (31-120 days)      4053
In Grace Period         2724
Charged Off             2493
Late (16-30 days)        970
Default                    2
Name: loan_status, dtype: int64

可以看到大部分都是正在还的期限当中或者已经完全付清。在本次评分卡当中，将除了current与fully paid以外的都设置为逾期。

In [11]:
#计算样本之比
(sum(df.loan_status == 'Current')+sum(df.loan_status == 'Fully Paid'))/df.shape[0]

0.9570650351994636

从结果中可以很明显看到数据极度不平衡。

In [12]:
#察看分类数据的情况值
df.select_dtypes(include=['O']).describe().T

count unique  \
id                              4      4   
term                       238636      2   
int_rate                   238636     75   
grade                      238636      7   
sub_grade                  238636     35   
emp_title                  218361  71700   
emp_length                 218740     11   
home_ownership             238636      4   
verification_status        238636      3   
issue_d                    238636      6   
loan_status                238636      7   
pymnt_plan                 238636      2   
purpose                    238636     13   
title                      238636     12   
zip_code                   238636    886   
addr_state                 238636     50   
earliest_cr_line           238636    661   
revol_util                 238329   1097   
initial_list_status        238636      2   
last_pymnt_d               238300     11   
next_pymnt_d               215724      3   
last_credit_pull_d         238635     13   
application_type           238636      2   
verification_status_joint   33434      3   
sec_app_earliest_cr_line    34418    602   
hardship_flag              238636      2   
hardship_type                 104      1   
hardship_reason               104      7   
hardship_status               104      3   
hardship_start_date           104      5   
hardship_end_date             104      5   
payment_plan_start_date       104      5   
hardship_loan_status          104      4   
disbursement_method        238636      2   
debt_settlement_flag       238636      2   
debt_settlement_flag_date     110      7   
settlement_status             110      3   
settlement_date               110      7   

                                                                        top  \
id                         Total amount funded in policy code 1: 1741781700   
term                                                              36 months   
int_rate                                                              5.31%   
grade                                                                     B   
sub_grade                                                                B5   
emp_title                                                           Teacher   
emp_length                                                        10+ years   
home_ownership                                                     MORTGAGE   
verification_status                                         Source Verified   
issue_d                                                            May-2018   
loan_status                                                         Current   
pymnt_plan                                                                n   
purpose                                                  debt_consolidation   
title                                                    Debt consolidation   
zip_code                                                              112xx   
addr_state                                                               CA   
earliest_cr_line                                                   Aug-2006   
revol_util                                                               0%   
initial_list_status                                                       w   
last_pymnt_d                                                       Nov-2018   
next_pymnt_d                                                       Dec-2018   
last_credit_pull_d                                                 Nov-2018   
application_type                                                 Individual   
verification_status_joint                                      Not Verified   
sec_app_earliest_cr_line                                           Aug-2006   
hardship_flag                                                             N   
hardship_type                               INTEREST ONLY-3 MONTHS DEFERRAL   
hardship_reason                                            NATURAL_DISASTER   
hardship_status     

发现：  
1.职业这一栏的种类有点多，不方便建模，需要进行察看。  
2.last_pymnt_d, next_pymnt_d, last_credit_pull_d, initial_list_status, pymnt_plan这几项均属于申请后的信息，与A卡无关，进行删除。  
3.application_type， disbursement_method, debt_settlement_flag这几项unique数少，数据分布极度不平衡，并且，对于了解数据集帮助不大，因此决定删除。  
4.zip_code与addr_state虽然属于贷前信息，但是按照州地区来进行分类，这对于不了解美国经济的我来说，不现实，决定删除。  
5.earliest_cr_line种类过多进行删除。

In [13]:
df.emp_title.value_counts()

Teacher                                    4295
Manager                                    4160
Owner                                      3830
Driver                                     2295
Registered Nurse                           1737
Sales                                      1613
Supervisor                                 1560
RN                                         1498
owner                                      1156
Office Manager                             1156
Project Manager                            1152
General Manager                             956
President                                   939
Director                                    918
Engineer                                    817
Nurse                                       715
Truck Driver                                661
manager                                     655
Operations Manager                          651
Accountant                                  626
teacher                                 

emp_title种类实在太多，并且难以分类，因此删除。

In [14]:
#对于变量类型为float的数据进行察看
df.select_dtypes(include = ['float64']).info(max_cols = 150)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 238640 entries, 0 to 238639
Data columns (total 107 columns):
member_id                                     0 non-null float64
loan_amnt                                     238636 non-null float64
funded_amnt                                   238636 non-null float64
funded_amnt_inv                               238636 non-null float64
installment                                   238636 non-null float64
annual_inc                                    238636 non-null float64
url                                           0 non-null float64
desc                                          0 non-null float64
dti                                           238048 non-null float64
delinq_2yrs                                   238636 non-null float64
inq_last_6mths                                238636 non-null float64
mths_since_last_delinq                        105569 non-null float64
mths_since_last_record                        31827 non-null fl

其中，loan_amnt, funded_amnt, funded_amnt_inv,int_rate, installment, out_prncp, out_prncp_inv, total_pymnt, total_pymnt_inv, total_rec_prncp, total_rec_int, total_rec_late_fee, last_pymnt_amnt这些变量属于申请后信息(贷中，贷后信息)进行删除。当然loan_amnt也应该属于申请信息，按照金额或许可以分类，本次分析不考虑。

In [15]:
#对于变量类型为float的数据进行察看
df.select_dtypes(include = ['float64']).head()

member_id  loan_amnt  funded_amnt  funded_amnt_inv  installment  \
0        NaN    10000.0      10000.0          10000.0       243.24   
1        NaN    10000.0      10000.0          10000.0       233.05   
2        NaN     1000.0       1000.0           1000.0        32.47   
3        NaN     2600.0       2600.0           2600.0        81.43   
4        NaN    10000.0      10000.0          10000.0       304.54   

   annual_inc  url  desc    dti  delinq_2yrs  inq_last_6mths  \
0     42000.0  NaN   NaN  27.06          0.0             0.0   
1     42000.0  NaN   NaN  24.69          0.0             0.0   
2     44000.0  NaN   NaN  16.37          1.0             1.0   
3     62000.0  NaN   NaN  19.61          0.0             0.0   
4    175000.0  NaN   NaN   4.94          0.0             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                    74.0       9.0      1.0   
1                    32.0                     NaN      17.0      0.0   
2                    13.0                     NaN       9.0      0.0   
3                     NaN                     NaN       8.0      0.0   
4                     NaN                     NaN      12.0      0.0   

   revol_bal  total_acc  out_prncp  out_prncp_inv  total_pymnt  \
0     6434.0       24.0    9079.27        9079.27      1862.16   
1      707.0       34.0    9161.42        9161.42      1664.14   
2     2748.0       23.0     829.04         829.04       226.13   
3    11776.0        9.0    2141.68        2141.68       567.71   
4     6885.0       25.0    8195.10        8195.10      2125.04   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1862.16           920.73         941.43                 0.0   
1          1664.14           838.58         825.56                 0.0   
2           226.13           170.96          55.17                 0.0   
3           567.71           458.32         109.39                 0.0   
4          2125.04          1804.90         320.14                 0.0   

   recoveries  collection_recovery_fee  last_pymnt_amnt  \
0         0.0                      0.0           243.24   
1         0.0                      0.0           233.05   
2         0.0                      0.0            32.47   
3         0.0                      0.0            81.43   
4         0.0                      0.0           304.54   

   collections_12_mths_ex_med  mths_since_last_major_derog  policy_code  \
0                         0.0                          NaN          1.0   
1                         0.0                         40.0          1.0   
2                         0.0                          NaN          1.0   
3                         0.0                          NaN          1.0   
4                         0.0                          NaN          1.0   

   annual_inc_joint  dti_joint  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0               NaN        NaN             0.0           0.0      31425.0   
1               NaN        NaN             0.0           0.0      93913.0   
2               NaN        NaN             0.0           0.0     173548.0   
3               NaN        NaN             0.0           0.0     241067.0   
4               NaN        NaN             0.0           0.0     318464.0   

   open_acc_6m  open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0          2.0          3.0          0.0          1.0                18.0   
1          0.0         15.0          0.0          0.0                54.0   
2          1.0          2.0          1.0          3.0                 3.0   
3          0.0          2.0          0.0          0.0                34.0   
4          0.0          2.0          0.0          3.0                16.0   

   total_bal_il  il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  \
0       24991.0     57.0          2.0          2.0      1186.0      51.0   
1       93206.0    116.0          0.0 

In [16]:
df.select_dtypes(include = ['float64']).describe().T

count           mean  \
member_id                                        0.0            NaN   
loan_amnt                                   238636.0   16016.908178   
funded_amnt                                 238636.0   16016.908178   
funded_amnt_inv                             238636.0   16013.282781   
installment                                 238636.0     466.073304   
annual_inc                                  238636.0   78609.597862   
url                                              0.0            NaN   
desc                                             0.0            NaN   
dti                                         238048.0      19.450565   
delinq_2yrs                                 238636.0       0.230037   
inq_last_6mths                              238636.0       0.453662   
mths_since_last_delinq                      105569.0      36.784814   
mths_since_last_record                       31827.0      81.009992   
open_acc                                    238636.0      11.413014   
pub_rec                                     238636.0       0.146801   
revol_bal                                   238636.0   15785.540027   
total_acc                                   238636.0      22.538443   
out_prncp                                   238636.0   12479.653472   
out_prncp_inv                               238636.0   12476.973353   
total_pymnt                                 238636.0    4427.707390   
total_pymnt_inv                             238636.0    4426.587860   
total_rec_prncp                             238636.0    3366.579956   
total_rec_int                               238636.0    1057.935467   
total_rec_late_fee                          238636.0       0.401324   
recoveries                                  238636.0       2.790644   
collection_recovery_fee                     238636.0       0.489836   
last_pymnt_amnt                             238636.0    1529.758439   
collections_12_mths_ex_med                  238636.0       0.017147   
mths_since_last_major_derog                  55823.0      46.129552   
policy_code                                 238636.0       1.000000   
annual_inc_joint                             34418.0  126019.332193   
dti_joint                                    34418.0      19.497580   
acc_now_delinq                              238636.0       0.000109   
tot_coll_amt                                238636.0     240.075140   
tot_cur_bal                                 238636.0  141832.407420   
open_acc_6m                                 238636.0       0.894404   
open_act_il                                 238636.0       2.665386   
open_il_12m                                 238636.0       0.666827   
open_il_24m                                 238636.0       1.506596   
mths_since_rcnt_il                          229315.0      21.615995   
total_bal_il                                238636.0   34735.182026   
il_util                                     198600.0      67.986405   
open_rv_12m                                 238636.0       1.252602   
open_rv_24m                                 238636.0       2.673360   
max_bal_bc                                  238636.0    5691.950770   
all_util                                    238573.0      53.831339   
total_rev_hi_lim                            238636.0   37591.630462   
inq_fi                                      238636.0       1.080009   
total_cu_tl                                 238636.0       1.460781   
inq_last_12m                                238636.0       1.994666   
acc_open_past_24mths                        238636.0       4.459139   
avg_cur_bal                                 238621.0   13652.621257   
bc_open_to_buy                              235386.0   14908.495701   
bc_util                                     235283.0      49.363118   
chargeoff_within_12_mths                    238636.0       0.006793   
delinq_amnt                                 238636.0       2.471857   


发现：  
1.member_id,url,desc,mths_since_last_delinq等字段数量较少(小于65%)需要删除。  
2.对于其他的数值型变量使用中位数来进行填充。 
3.发现delinq_2yrs等变量至少有75%或者以上的数据是相同的，这可能是由于数据是不平衡的，需要查看。
4.policy_code，num_tl_120dpd_2m等变量标准差为零，需要删除。

**需要处理的问题归纳**

需要处理的问题汇总：  
1.有4行的数据毫无意义需要删除。    
2.对于困难情况以及其他对于其他的数据量小于65%的字段进行删除。    
3.对于term,grade,sub_grade,issue_d,title均属于贷后数据因此决定删除。  
4.zip_code,addr_state,earliest_cr_line,emp_title种类过多进行字段进行删除。  
5.application_type， disbursement_method, debt_settlement_flag这几项unique数少，数据分布极度不平衡，并且，对于了解数据集帮助不大，因此决定删除。   
6.last_pymnt_d, next_pymnt_d, last_credit_pull_d, initial_list_status, pymnt_plan这几项均属于申请后的信息，与A卡无关，进行删除。  
7.对于loan_amnt, funded_amnt, funded_amnt_inv, int_rate, installment, out_prncp, out_prncp_inv, total_pymnt, total_pymnt_inv, total_rec_prncp, total_rec_int,total_rec_late_fee,last_pymnt_amnt 这些变量进行删除。 
8.emp_length这个字段的缺失值较少的字段的缺失值用unknown来进行填充。   
9.int_rate, revol_util这两个本应该为数字的字段这里是用字符串表示的，为了之后的计算方便，需要更改类型。(由于发现int_rate为贷后字段，只对revol_util进行操作)    
10.对于缺失较少的字段用中位数进行填充。  
11.发现delinq_2yrs等变量至少有75%或者以上的数据是相同的，进行删除。  
12.policy_code，num_tl_120dpd_2m等变量标准差为零，需要删除。  
13.将loan_status中的类别转化为0和1(1代表不良)。 

**数据清洗**

In [17]:
#复制数据集
df_clean = df.copy()

1.有4行的数据毫无意义需要删除。

In [18]:
#对于无意义的4行进行删除
df_clean = df_clean.loc[df.id.isnull()]

In [19]:
#确认
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 238636 entries, 0 to 238637
Columns: 145 entries, id to settlement_term
dtypes: float64(107), object(38)
memory usage: 265.8+ MB


2.对于困难情况以及其他对于其他的数据量小于65%的字段进行删除。 

In [20]:
#对于与困难情况相关的字段，以及数据量小于65%的字段进行删除
thr = 0.65
df_clean = df_clean.dropna(thresh = len(df_clean)*thr, axis = 1)
df_clean = df_clean.drop(columns = 'hardship_flag', axis = 1)

3.对于term,grade,sub_grade,issue_d,title均属于贷后数据因此决定删除。 

In [21]:
#对于相应字段进行删除
col = ['term','grade','sub_grade','issue_d','title']
df_clean.drop(columns = col, axis = 1, inplace = True)

4.zip_code,addr_state,earliest_cr_line,emp_title种类过多进行字段进行删除。 

In [22]:
#对于相应字段进行删除
col = ['zip_code', 'addr_state','earliest_cr_line','emp_title']
df_clean.drop(columns = col, axis = 1, inplace = True)

application_type， disbursement_method, debt_settlement_flag这几项unique数少，数据分布极度不平衡，并且，对于了解数据集帮助不大，因此决定删除。

In [23]:
#对于相应字段进行删除
col = ['application_type', 'disbursement_method', 'debt_settlement_flag']
df_clean.drop(columns = col, axis = 1, inplace = True)

6.last_pymnt_d, next_pymnt_d, last_credit_pull_d, initial_list_status, pymnt_plan这几项均属于申请后的信息，与A卡无关，进行删除。  

In [24]:
#对于相应字段进行删除
col = ['last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d', 'initial_list_status', 'pymnt_plan']
df_clean.drop(columns = col, axis = 1, inplace = True)

7.对于loan_amnt, funded_amnt, funded_amnt_inv, int_rate, installment, out_prncp, out_prncp_inv, total_pymnt, total_pymnt_inv, total_rec_prncp, total_rec_int,total_rec_late_fee,last_pymnt_amnt 这些变量进行删除。 

In [25]:
col = ['loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'int_rate', 'installment', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int','total_rec_late_fee','last_pymnt_amnt']
df_clean.drop(columns = col, axis = 1, inplace = True)

8.emp_length这个字段的缺失值较少的字段的缺失值用unknown来进行填充。 

In [26]:
#对于emp_title, emp_length的缺失值进行填充
df_clean.emp_length.loc[df_clean.emp_length.isnull()] = 'unknown'

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\indexing.py:189: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


In [27]:
#确认
print(sum(df_clean.emp_length.isnull()))

0


9.int_rate, revol_util这两个本应该为数字的字段这里是用字符串表示的，为了之后的计算方便，需要更改类型。(由于发现int_rate为贷后字段，只对revol_util进行操作) 

In [28]:
#用正则表达式提取数字部分,并且赋值
df_clean.revol_util = df_clean.revol_util.str.extract(r'(\d+\.\d+)')
df_clean.revol_util = df_clean.revol_util.astype('float64')

In [29]:
#确认
print(df_clean.revol_util.dtype)

float64


10.对于缺失较少的字段用中位数进行填充。

In [30]:
#用中位数进行填充
for col in df_clean.columns:
    if df_clean[col].isnull().any():
        df_clean[col].fillna(df_clean[col].median(),inplace=True)

In [31]:
#确认
for col in df_clean.columns:
    if df_clean[col].isnull().any():
        print('失败')

11.发现delinq_2yrs等变量至少有75%或者以上的数据是相同的，进行删除。

In [32]:
#删除信息量过少的数据(这里按照第三四分位叔等于最小值来判断)
col = []
for i in df_clean.select_dtypes(include=['float64']).columns:
    if df_clean[i].quantile(.75) == df_clean[i].min():
        col.append(i)
col

['delinq_2yrs',
 'pub_rec',
 'recoveries',
 'collection_recovery_fee',
 'collections_12_mths_ex_med',
 'policy_code',
 'acc_now_delinq',
 'tot_coll_amt',
 'chargeoff_within_12_mths',
 'delinq_amnt',
 'num_accts_ever_120_pd',
 'num_tl_120dpd_2m',
 'num_tl_30dpd',
 'num_tl_90g_dpd_24m',
 'pub_rec_bankruptcies',
 'tax_liens']

In [33]:
df_clean.drop(columns=col,inplace=True)

12.policy_code，num_tl_120dpd_2m等变量标准差为零，需要删除。(发现在11中已经完成)

13.将loan_status中的类别转化为0和1(1代表不良)。

In [34]:
df_clean = df_clean.reset_index(drop = True)
count = 0

df_clean['class'] = df_clean.loan_status.apply(lambda x: '0' if x in ('Current', 'Fully Paid') else '1')

In [35]:
#确认
df_clean['class'].value_counts()

0    228394
1     10242
Name: class, dtype: int64

In [36]:
#删除loan_status字段
df_clean.drop(columns = 'loan_status', inplace = True)

In [37]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 238636 entries, 0 to 238635
Data columns (total 55 columns):
emp_length                    238636 non-null object
home_ownership                238636 non-null object
annual_inc                    238636 non-null float64
verification_status           238636 non-null object
purpose                       238636 non-null object
dti                           238636 non-null float64
inq_last_6mths                238636 non-null float64
open_acc                      238636 non-null float64
revol_bal                     238636 non-null float64
revol_util                    238636 non-null float64
total_acc                     238636 non-null float64
tot_cur_bal                   238636 non-null float64
open_acc_6m                   238636 non-null float64
open_act_il                   238636 non-null float64
open_il_12m                   238636 non-null float64
open_il_24m                   238636 non-null float64
mths_since_rcnt_il           

In [38]:
#写入
df_clean.to_csv('../data/rough_clean_data.csv',index = False)